fluture.js

```diff
@@ -376,544 +404,927 @@
-    });
+  SafeFuture.prototype.toString = function SafeFuture$toString(){
+    return `Future(${showf(this._computation)})`;
   }
 
-  function Future$recur(f){
-    check$recur(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$chain$fork(rej, res){
-      return _this._f(rej, function Future$chain$res(x){
-        const m = f(x);
-        check$recur$f(m, f, x);
-        m._f(rej, res);
-      });
-    });
+  //----------
+
+  function ChainRec(iterate, init){
+    this._iterate = iterate;
+    this._init = init;
   }
 
-  function Future$chainRej(f){
-    check$chainRej(this, f);
+  ChainRec.prototype = Object.create(Future.prototype);
+
+  ChainRec.prototype._f = function ChainRec$fork(rej, res){
     const _this = this;
-    return new FutureClass(false, function Future$chainRej$fork(rej, res){
-      let cancel;
-      const r = _this._f(function Future$chainRej$rej(x){
-        const m = f(x);
-        check$chainRej$f(m, f, x);
-        cancel = m._f(rej, res);
-      }, res);
-      return cancel ? cancel : (cancel = r, function Future$chainRej$cancel(){ cancel() });
-    });
+    let cancel = noop, i = 0;
+    (function Future$chainRec$recur(state){
+      let isSync = null;
+      function Future$chainRec$res(it){
+        check$chainRec$it(it, i);
+        i = i + 1;
+        if(isSync === null){
+          isSync = true;
+          state = it;
+        }else{
+          Future$chainRec$recur(it);
+        }
+      }
+      while(!state.done){
+        isSync = null;
+        const m = _this._iterate(Next, Done, state.value);
+        check$chainRec$f(m, _this._iterate, i, state.value);
+        cancel = m._f(rej, Future$chainRec$res);
+        if(isSync === true){
+          continue;
+        }else{
+          isSync = false;
+          return;
+        }
+      }
+      res(state.value);
+    }(Next(_this._init)));
+    return function Future$chainRec$cancel(){ cancel() };
   }
 
-  function Future$map(f){
-    check$map(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$map$fork(rej, res){
-      return _this._f(rej, function Future$map$res(x){
-        res(f(x));
-      });
-    });
+  ChainRec.prototype.toString = function ChainRec$toString(){
+    return `Future.chainRec(${showf(this._iterate)}, ${show(this._init)})`;
   }
 
-  function Future$mapRej(f){
-    check$mapRej(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$mapRej$fork(rej, res){
-      return _this._f(function Future$mapRej$rej(x){
-        rej(f(x));
-      }, res);
-    });
+  //----------
+
+  function CachedFuture(pure){
+    this._pure = pure;
+    this._cancel = noop;
+    this._queue = [];
+    this._queued = 0;
+    this._value = null;
+    this._state = 0;
   }
 
-  function Future$bimap(f, g){
-    check$bimap(this, f, g);
+  CachedFuture.STATE = {
+    0: 'cold',
```

Comment:
Why not flip these and have the keys be the names so the code that utilizes the state can be easier to understand?